include/rmm/mr/device/sam_headroom_resource_adaptor.hpp

```diff
@@ -0,0 +1,165 @@
+ * the allocation would not eat in to the headroom. Since doing this check on every allocation can
+ * be expensive, checking whether to set the preferred location is only done above the specified
+ * allocation threshold size.
+ *
+ * Note that if threshold size is non-zero, then an application making many small allocations can
+ * eat into the headroom, and run out of memory for other CUDA calls, even when using this resource
+ * adaptor.
+ *
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation. Must be
+ *                  `system_memory_resource`.
+ */
+template <typename Upstream>
+class sam_headroom_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a headroom adaptor using `upstream` to satisfy allocation requests.
+   *
+   * @param upstream The resource used for allocating/deallocating device memory. Must be
+   *                 `system_memory_resource`.
+   * @param headroom Size of the reserved GPU memory as headroom
+   * @param threshold Size of the allocation above which to check for headroom
+   */
+  explicit sam_headroom_resource_adaptor(Upstream* upstream,
+                                         std::size_t headroom,
+                                         std::size_t threshold)
+    : upstream_{upstream}, headroom_{headroom}, threshold_{threshold}
+  {
+    static_assert(std::is_same_v<system_memory_resource, Upstream>,
+                  "Upstream must be rmm::mr::system_memory_resource");
+  }
+
+  sam_headroom_resource_adaptor()                                                = delete;
+  ~sam_headroom_resource_adaptor() override                                      = default;
+  sam_headroom_resource_adaptor(sam_headroom_resource_adaptor const&)            = delete;
+  sam_headroom_resource_adaptor(sam_headroom_resource_adaptor&&)                 = delete;
+  sam_headroom_resource_adaptor& operator=(sam_headroom_resource_adaptor const&) = delete;
+  sam_headroom_resource_adaptor& operator=(sam_headroom_resource_adaptor&&)      = delete;
+
+  /**
+   * @briefreturn{rmm::device_async_resource_ref to the upstream resource}
+   */
+  [[nodiscard]] rmm::device_async_resource_ref get_upstream_resource() const noexcept
+  {
+    return upstream_;
+  }
+
+  /**
+   * @briefreturn{Upstream* to the upstream memory resource}
+   */
+  [[nodiscard]] Upstream* get_upstream() const noexcept { return upstream_; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * The stream argument is ignored.
+   *
+   * @param bytes The size of the allocation
+   * @param stream This argument is ignored
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, [[maybe_unused]] cuda_stream_view stream) override
+  {
+    auto const aligned{rmm::align_up(bytes, rmm::CUDA_ALLOCATION_ALIGNMENT)};
```

Comment:
I went back and did some experiments with both HMM and ATS, looks like we don't actually need to do additional alignments. Removed.

Reply:
I got more recent advice from the CUDA team and they suggested we should use CUDA alignment if it will be accessed on the device. If nothing else, for cache line alignment this is important.

Reply:
How do we do this alignment? Do we need to add some padding as in `aligned_host_allocate`?

Reply:
Changed to do host MR style alignment, but with CUDA alignment. PTAL